Intermittent test failure due to concurrent modification of enumerable

I have no idea what causes it, but I'm putting it here so that maybe I'll take a look at it something

```
Painting map RenderMapPrototype(Dev) failed due to an internal exception:
System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.Collections.Generic.HashSet`1.Enumerator.MoveNext()
   at Robust.Shared.GameObjects.TransformChildrenEnumerator.MoveNext(EntityUid& child) in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.Shared/GameObjects/Components/Transform/TransformComponent.cs:line 570
   at Robust.Client.GameStates.ClientGameStateManager.ProcessDeletions(ReadOnlySpan`1 delSpan, EntityQuery`1 xforms, EntityQuery`1 metas, SharedTransformSystem xformSys) in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.Client/GameStates/ClientGameStateManager.cs:line 1241
   at Robust.Client.GameStates.ClientGameStateManager.ApplyEntityStates(GameState curState, GameState nextState) in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.Client/GameStates/ClientGameStateManager.cs:line 954
   at Robust.Client.GameStates.ClientGameStateManager.ApplyGameState(GameState curState, GameState nextState) in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.Client/GameStates/ClientGameStateManager.cs:line 782
   at Robust.Client.GameStates.ClientGameStateManager.ApplyGameState() in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.Client/GameStates/ClientGameStateManager.cs:line 409
   at Robust.Client.GameController.Tick(FrameEventArgs frameEventArgs) in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.Client/GameController/GameController.cs:line 618
   at Robust.Client.GameController.<>c__DisplayClass75_0.<StartupContinue>b__7(Object sender, FrameEventArgs args) in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.Client/GameController/GameController.cs:line 291
   at Robust.UnitTesting.RobustIntegrationTest.IntegrationGameLoop.SingleThreadRunUntilEmpty() in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.UnitTesting/RobustIntegrationTest.cs:line 1163
   at Robust.UnitTesting.RobustIntegrationTest.IntegrationGameLoop.Run() in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.UnitTesting/RobustIntegrationTest.cs:line 1140
   at Robust.UnitTesting.RobustIntegrationTest.ClientIntegrationInstance.ThreadMain() in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.UnitTesting/RobustIntegrationTest.cs:line 951
--- End of stack trace from previous location ---
   at Robust.UnitTesting.RobustIntegrationTest.IntegrationInstance.WaitIdleImplAsync(Boolean throwOnUnhandled, CancellationToken cancellationToken) in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.UnitTesting/RobustIntegrationTest.cs:line 483
   at Robust.UnitTesting.RobustIntegrationTest.IntegrationInstance.WaitRunTicks(Int32 ticks) in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.UnitTesting/RobustIntegrationTest.cs:line 570
   at Robust.UnitTesting.Pool.TestPair`2.RunTicksSync(Int32 ticks) in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.UnitTesting/Pool/TestPair.Helpers.cs:line 185
   at Content.MapRenderer.Painters.MapPainter.Paint()+MoveNext() in /home/runner/work/moff-station-14/moff-station-14/Content.MapRenderer/Painters/MapPainter.cs:line 188
   at Content.MapRenderer.Painters.MapPainter.Paint()+System.Threading.Tasks.Sources.IValueTaskSource<System.Boolean>.GetResult()
   at Content.MapRenderer.Program.Run(CommandLineArguments arguments, List`1 toRender, ExternalTestContext testContext) in /home/runner/work/moff-station-14/moff-station-14/Content.MapRenderer/Program.cs:line 212
   at Content.MapRenderer.Program.Run(CommandLineArguments arguments, List`1 toRender, ExternalTestContext testContext) in /home/runner/work/moff-station-14/moff-station-14/Content.MapRenderer/Program.cs:line 212
DisposeAsync: Dirty return of pair 1 started
DisposeAsync: Test gave back pair 1 in 188.6054 ms
DisposeAsync: Disposed pair 1 in 0.4494 ms
Unhandled exception. NUnit.Framework.MultipleAssertException: CLIENT: 9.028s [ERRO] runtime: Caught exception in "ClientGameStateManager.ApplyEntityStates caught exception while deleting entities!" Exception: System.InvalidOperationException: Collection was modified; enumeration operation may not execute.
   at System.Collections.Generic.HashSet`1.Enumerator.MoveNext()
   at Robust.Shared.GameObjects.TransformChildrenEnumerator.MoveNext(EntityUid& child) in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.Shared/GameObjects/Components/Transform/TransformComponent.cs:line 570
   at Robust.Client.GameStates.ClientGameStateManager.ProcessDeletions(ReadOnlySpan`1 delSpan, EntityQuery`1 xforms, EntityQuery`1 metas, SharedTransformSystem xformSys) in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.Client/GameStates/ClientGameStateManager.cs:line 1241
   at Robust.Client.GameStates.ClientGameStateManager.ApplyEntityStates(GameState curState, GameState nextState) in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.Client/GameStates/ClientGameStateManager.cs:line 954
   at NUnit.Framework.Assert.AssertionScope.Dispose()
   at Robust.UnitTesting.Pool.TestPair`2.ReportErrorLogs() in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.UnitTesting/Pool/TestPair.Recycle.cs:line 34
   at Robust.UnitTesting.Pool.TestPair`2.ReportErrorLogs() in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.UnitTesting/Pool/TestPair.Recycle.cs:line 22
   at Robust.UnitTesting.Pool.TestPair`2.OnDirtyDispose() in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.UnitTesting/Pool/TestPair.Recycle.cs:line 45
   at Robust.UnitTesting.Pool.TestPair`2.DisposeAsync() in /home/runner/work/moff-station-14/moff-station-14/RobustToolbox/Robust.UnitTesting/Pool/TestPair.Recycle.cs:line 150
   at Content.MapRenderer.Painters.MapPainter.DisposeAsync() in /home/runner/work/moff-station-14/moff-station-14/Content.MapRenderer/Painters/MapPainter.cs:line 245
   at Content.MapRenderer.Program.Run(CommandLineArguments arguments, List`1 toRender, ExternalTestContext testContext) in /home/runner/work/moff-station-14/moff-station-14/Content.MapRenderer/Program.cs:line 266
   at Content.MapRenderer.Program.Main(String[] args) in /home/runner/work/moff-station-14/moff-station-14/Content.MapRenderer/Program.cs:line 180
   at Content.MapRenderer.Program.<Main>(String[] args)
Error: Process completed with exit code 134.
```

--- a/Content.Client/Content.Client.csproj
+++ b/Content.Client/Content.Client.csproj
@@ -21,11 +21,6 @@
   <ItemGroup>
     <ProjectReference Include="..\Content.Shared\Content.Shared.csproj" />
   </ItemGroup>
-
-  <ItemGroup>
-    <Folder Include="_Moffstation\Robotics\" />
-    <Folder Include="_Moffstation\Silicons\" />
-  </ItemGroup>
   <Import Project="..\RobustToolbox\MSBuild\Robust.Properties.targets" />
   <Import Project="..\RobustToolbox\MSBuild\XamlIL.targets" />
 </Project>

--- a/Content.Client/_Moffstation/NightVision/DirtyNightVisionComponent.cs
+++ b/Content.Client/_Moffstation/NightVision/DirtyNightVisionComponent.cs
@@ -1,0 +1,6 @@
+﻿namespace Content.Client._Moffstation.NightVision;
+
+/// This marker component exists only to indicate that its owning entity needs to have NightVision applied. Does nothing
+/// if on an entity without <see cref="Content.Shared._Moffstation.NightVision"/>
+[RegisterComponent]
+public sealed partial class DirtyNightVisionComponent : Component;

--- a/Content.Client/_Moffstation/NightVision/NightVisionSystem.cs
+++ b/Content.Client/_Moffstation/NightVision/NightVisionSystem.cs
@@ -81,7 +81,7 @@ public sealed class NightVisionSystem : EntitySystem
     private void RemoveEffect(Entity<NightVisionComponent> entity)
     {
         _overlayMan.RemoveOverlay<NightVisionOverlay>();
-        QueueDel(entity.Comp.Effect);
+        PredictedQueueDel(entity.Comp.Effect);
         entity.Comp.Effect = null;
         RemCompDeferred<DirtyNightVisionComponent>(entity);
     }

--- a/Content.Client/_Moffstation/NightVision/NightVisionSystem.cs
+++ b/Content.Client/_Moffstation/NightVision/NightVisionSystem.cs
@@ -1,25 +1,23 @@
 using Content.Client._Starlight.Overlay;
-using Content.Shared._Moffstation.Overlay.Components;
+using Content.Shared._Moffstation.NightVision;
 using Content.Shared.Flash;
-using Content.Shared.Flash.Components;
-using Content.Shared.Inventory;
 using Robust.Client.GameObjects;
 using Robust.Client.Graphics;
 using Robust.Client.Player;
 using Robust.Shared.Player;
-using Robust.Shared.Prototypes;
+using Robust.Shared.Timing;
 
-namespace Content.Client._Moffstation.Overlay.Systems;
+namespace Content.Client._Moffstation.NightVision;
 
-/// <summary>
 /// This system implements the behavior of <see cref="NightVisionComponent"/>.
-/// </summary>
 public sealed class NightVisionSystem : EntitySystem
 {
+    [Dependency] private readonly SharedFlashSystem _flash = default!;
+    [Dependency] private readonly IGameTiming _gameTiming = default!;
     [Dependency] private readonly IPlayerManager _player = default!;
     [Dependency] private readonly IOverlayManager _overlayMan = default!;
     [Dependency] private readonly TransformSystem _xformSys = default!;
-    [Dependency] private readonly SharedFlashSystem _flash = default!;
+
     public override void Initialize()
     {
         base.Initialize();
@@ -35,6 +33,10 @@ public sealed class NightVisionSystem : EntitySystem
 
     private void OnFlashImmunityChanged(Entity<NightVisionComponent> ent, ref FlashImmunityChangedEvent args)
     {
+        // Prevent effect flickering due to rollbacks.
+        if (!_gameTiming.IsFirstTimePredicted)
+            return;
+
         if (args.FlashImmune)
         {
             RemoveEffect(ent);
@@ -72,16 +74,35 @@ public sealed class NightVisionSystem : EntitySystem
             _flash.IsFlashImmune(entity))
             return;
 
-        _overlayMan.AddOverlay(new NightVisionOverlay());
-        var effect = SpawnAttachedTo(entity.Comp.EffectPrototype, Transform(entity).Coordinates);
-        _xformSys.SetParent(effect, entity);
-        entity.Comp.Effect = effect;
+        // "Enqueue" activation of night vision.
+        EnsureComp<DirtyNightVisionComponent>(entity);
     }
 
     private void RemoveEffect(Entity<NightVisionComponent> entity)
     {
         _overlayMan.RemoveOverlay<NightVisionOverlay>();
-        PredictedQueueDel(entity.Comp.Effect);
+        QueueDel(entity.Comp.Effect);
         entity.Comp.Effect = null;
+        RemCompDeferred<DirtyNightVisionComponent>(entity);
+    }
+
+    public override void Update(float frameTime)
+    {
+        // We use the marker+update pattern for changing here due a pernicious intermittent issue where night vision
+        // would be turned on while an entity was being deleted as part of round-end which would cause
+        // https://github.com/moff-station/moff-station-14/issues/1293 (or something like it) due to adding entities
+        // during state application, which iterates over existing entities.
+        var q = AllEntityQuery<NightVisionComponent, DirtyNightVisionComponent>();
+        while (q.MoveNext(out var ent, out var comp, out _))
+        {
+            if (comp.Effect != null)
+                continue;
+
+            _overlayMan.AddOverlay(new NightVisionOverlay());
+            var effect = SpawnAttachedTo(comp.EffectPrototype, Transform(ent).Coordinates);
+            _xformSys.SetParent(effect, ent);
+            comp.Effect = effect;
+            RemCompDeferred<DirtyNightVisionComponent>(ent);
+        }
     }
 }

--- a/Content.Shared/_Moffstation/Clothing/ModularHud/Systems/SharedModularHudSystem.cs
+++ b/Content.Shared/_Moffstation/Clothing/ModularHud/Systems/SharedModularHudSystem.cs
@@ -8,7 +8,6 @@ using Content.Shared.DoAfter;
 using Content.Shared.Examine;
 using Content.Shared.Eye.Blinding.Systems;
 using Content.Shared.Flash;
-using Content.Shared.Foldable;
 using Content.Shared.Hands.EntitySystems;
 using Content.Shared.IdentityManagement.Components;
 using Content.Shared.Interaction;

--- a/Content.Shared/_Moffstation/NightVision/NightVisionComponent.cs
+++ b/Content.Shared/_Moffstation/NightVision/NightVisionComponent.cs
@@ -1,7 +1,7 @@
 using Robust.Shared.GameStates;
 using Robust.Shared.Prototypes;
 
-namespace Content.Shared._Moffstation.Overlay.Components;
+namespace Content.Shared._Moffstation.NightVision;
 
 /// When this component is on a player-controlled entity, it applies a night vision visual effect
 /// (<see cref="Content.Client._Starlight.Overlay.NightVisionOverlay"/>) and creates a
@@ -12,7 +12,7 @@ public sealed partial class NightVisionComponent : Component
     [DataField, AutoNetworkedField]
     public EntProtoId EffectPrototype = "EffectNightVision";
 
-    /// The instance of <see cref="EffectPrototype"/>..
+    /// The instance of <see cref="EffectPrototype"/>.
     [ViewVariables]
     public EntityUid? Effect;
 }

